File: examples/thermal/ctpq/cpu_none.py
"""thermal | cTPQ | CPU (OpenMP) | full Hilbert (no symmetry)

Canonical Thermal Pure Quantum (Krylov imaginary-time). Twin: ``examples/thermal/ctpq/cpu_none.cpp``.

Requires: (no special requirements; runs on the default CPU build)
"""
from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from _shared.common import heisenberg_chain, rank0_print

import qed

N = 8
H = heisenberg_chain(N, pbc=True)

result = qed.thermal(
    H,
    method="cTPQ",
    T_min=0.1,
    T_max=10.0,
    num_T=8,
    num_samples=4,
    random_seed=0,
    device="cpu",
    verbose=False,
)

T  = result.temperatures
E  = result.energy
Cv = result.specific_heat
mid = len(T) // 2

rank0_print(f"gs_E    = {result.ground_state_energy:.4f}")
rank0_print(f"T[0]    = {T[0]:.4f}  E = {E[0]:.4f}  Cv = {Cv[0]:.4f}")
rank0_print(f"T[mid]  = {T[mid]:.4f}  E = {E[mid]:.4f}  Cv = {Cv[mid]:.4f}")
rank0_print(f"T[-1]   = {T[-1]:.4f}  E = {E[-1]:.4f}  Cv = {Cv[-1]:.4f}")

# === Expected output (deterministic; captured on the CI reference runner) ===
# gs_E    = -3.6511
# T[0]    = 0.1000  E = -3.6455  Cv = 0.2903
# T[mid]  = 5.7571  E = 0.5193  Cv = 0.0241
# T[-1]   = 10.0000  E = 0.6143  Cv = 0.0072
# ===========================================================================

File: examples/spectral/ground_state_dssf/cpu_none.cpp
// =============================================================================
// examples/spectral/ground_state_dssf/cpu_none.cpp
//
// spectral | ground-state DSSF | CPU (OpenMP) | full Hilbert (no symmetry)
//
// dynamical structure factor S_zz(omega) at three omega points. Twin: examples/spectral/ground_state_dssf/cpu_none.py
//
// Requires: (no special requirements; runs on the default CPU build)
// =============================================================================

#include "../../_shared/common.h"

#include <ed/api.h>

#include <iomanip>
#include <iostream>
#include <vector>

int main(int argc, char** argv) {
    auto guard = ed_example::mpi_guard(argc, argv);
    constexpr std::uint64_t N = 8;
    (void)guard;

    auto H_op = ed_example::heisenberg_chain(N, /*pbc=*/true);
    auto obs = std::make_unique<Operator>(N, /*spin=*/0.5f);
    // Staggered S_z(q=pi) = sum_i (-1)^i S_z_i. Non-trivially mixes with H
    // (S_z_total is a conserved charge and would make the DSSF identically
    // zero -- using the staggered combination produces a meaningful spectrum).
    for (std::uint64_t i = 0; i < N; ++i) {
        const double sign = (i % 2 == 0) ? 1.0 : -1.0;
        obs->addOneBodyTerm(2, i, std::complex<double>(sign, 0.0)); // (-1)^i Sz
    }

    auto h_spec   = ed_example::in_memory_spec(std::move(H_op), N);
    auto obs_op   = ed::make_operator(ed_example::in_memory_spec(std::move(obs), N));
    std::vector<const ed::LinearOperator*> observables = { obs_op.get() };

    ed::api::SpectralOptions opts;
    opts.method     = "ground_state_cf";
    opts.device     = "cpu";
    opts.omega_min  = -5.0;
    opts.omega_max  =  5.0;
    opts.num_omega  = 11;
    opts.eta        = 0.1;
    opts.krylov_dim = 80;

    auto result = ed::api::spectral(std::move(h_spec), observables, opts);

    const std::size_t mid = result.omega.size() / 2;
    std::cout << std::fixed << std::setprecision(6);
    ed_example::rank0_print("S(w=-5.0) = ", result.S_real.front(), "\n");
    ed_example::rank0_print("S(w= 0.0) = ", result.S_real[mid],   "\n");
    ed_example::rank0_print("S(w= 5.0) = ", result.S_real.back(),  "\n");

    // === Expected output (deterministic; captured on the CI reference runner) ===
    // S(w=-5.0) = 0.006872
    // S(w= 0.0) = 0.703124
    // S(w= 5.0) = 0.013104
    // ===========================================================================
    return 0;
}
